### PR TITLE
CSI support

### PR DESCRIPTION
**Pull request by niusmallnan**
Kubernetes 1.10+ is recommended, if you want to use RKE, please refer to following:
```
services:
  kube-api:
    extra_args:
      allow-privileged: true
      feature-gates: MountPropagation=true

  kubelet:
    extra_args:
      allow-privileged: true
      feature-gates: MountPropagation=true
```

About deploying longhorn CSI plugin, please refer to [longhorn-csi](https://github.com/niusmallnan/longhorn-csi) : 
```
# deploy csi
kubectl create -f deploy/csi/

# an example
kubectl create -f example/

# check logs
kubectl -n longhorn-system  logs -f --tail=20  pod/<csi-longhorn-plugin-pod> -c csi-longhorn-plugin
```

TODO list:
1. Automation and manual tests
2. CSI auto-deployment. Manual deployment is more convenient for debugging, will make another PR for auto-deployment.

Differences from the design document:
1. Driver name should be consistent with CSI sock dirname.
2. The List_Volumes func in the Controller server is not implemented. It is NOT REQUIRED for a controller plugin to implement the LIST_VOLUMES capability if it supports the CREATE_DELETE_VOLUME capability.

### CODE REVIEW

**Comment by yasker** [comment]
BTW, if we need extra parameters to make it work with RKE, we should try to make it default in RKE.

**Comment by niusmallnan** [comment]
@yasker PR updated and did the following tests:
1. Deploy on RKE and GKE
2. Manually create and delete CSI volumes
3. Tested disconnect one host when volume attached. The pod should be migrated by Kubernetes and Longhorn volume should be attached to it automatically.

Tried to run longhorn-manager integration tests but failed:
```
kubectl logs -f po/longhorn-test-nfs
* Starting Ganesha-NFS
28/05/2018 06:45:57 : epoch 5b0ba5a5 : longhorn-test-nfs : ganesha.nfsd-6[main] main :MAIN :EVENT :ganesha.nfsd Starting: Ganesha Version /builddir/build/BUILD/nfs-ganesha-2.3.2/src, built at May  4 2016 05:09:13 on
28/05/2018 06:45:57 : epoch 5b0ba5a5 : longhorn-test-nfs : ganesha.nfsd-6[main] nfs_set_param_from_conf :NFS STARTUP :EVENT :Configuration file successfully parsed
28/05/2018 06:45:57 : epoch 5b0ba5a5 : longhorn-test-nfs : ganesha.nfsd-6[main] init_server_pkgs :NFS STARTUP :EVENT :Initializing ID Mapper.
28/05/2018 06:45:57 : epoch 5b0ba5a5 : longhorn-test-nfs : ganesha.nfsd-6[main] init_server_pkgs :NFS STARTUP :EVENT :ID Mapper successfully initialized.
28/05/2018 06:45:57 : epoch 5b0ba5a5 : longhorn-test-nfs : ganesha.nfsd-6[main] export_defaults_display :CONFIG :EVENT :DEFAULTS Export Defaults (root_squash   , ----, 34-, UDP, TCP, ----, No Manage_Gids, -- Deleg, anon_uid=    -2, anon_gid=    -2, none, sys)
28/05/2018 06:45:57 : epoch 5b0ba5a5 : longhorn-test-nfs : ganesha.nfsd-6[main] export_defaults_display :CONFIG :EVENT :RESULT Export Defaults (root_squash   , ----, 34-, UDP, TCP, ----, No Manage_Gids, -- Deleg, anon_uid=    -2, anon_gid=    -2, none, sys)
28/05/2018 06:45:57 : epoch 5b0ba5a5 : longhorn-test-nfs : ganesha.nfsd-6[main] lower_my_caps :NFS STARTUP :EVENT :CAP_SYS_RESOURCE was successfully removed for proper quota management in FSAL
28/05/2018 06:45:57 : epoch 5b0ba5a5 : longhorn-test-nfs : ganesha.nfsd-6[main] lower_my_caps :NFS STARTUP :EVENT :currenty set capabilities are: = cap_chown,cap_dac_override,cap_dac_read_search,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_net_bind_service,cap_net_raw,cap_sys_chroot,cap_sys_admin,cap_mknod,cap_audit_write,cap_setfcap+eip
28/05/2018 06:45:57 : epoch 5b0ba5a5 : longhorn-test-nfs : ganesha.nfsd-6[main] vfs_lookup_path :FSAL :CRIT :Could not get handle for path /opt/backupstore, error Operation not supported
28/05/2018 06:45:57 : epoch 5b0ba5a5 : longhorn-test-nfs : ganesha.nfsd-6[main] init_export_root :EXPORT :CRIT :Lookup failed on path, ExportId=14 Path=/opt/backupstore FSAL_ERROR=(Operation not supported,95)
28/05/2018 06:45:57 : epoch 5b0ba5a5 : longhorn-test-nfs : ganesha.nfsd-6[main] nfs_Init_svc :DISP :CRIT :Cannot acquire credentials for principal nfs
28/05/2018 06:45:57 : epoch 5b0ba5a5 : longhorn-test-nfs : ganesha.nfsd-6[main] nfs_Init_admin_thread :NFS CB :EVENT :Admin thread initialized
bash: line 1:     6 Segmentation fault      (core dumped) /opt/start_nfs.sh
```

**Comment by yasker** [comment]
@niusmallnan That userspace Ganesha NFS server has issues with the certain version of Docker. I am waiting for @JacieChao to update it with minio and S3 support, then we can ask user to use minio for backup feature, and switch back basic test with kernel based nfs server.

For now, use this yaml file instead. But you will need to install `nfs-kernel-server` on every host first.
```
apiVersion: v1
kind: Pod
metadata:
  name: longhorn-test-nfs
  labels:
    app: longhorn-test-nfs
spec:
  containers:
  - name: longhorn-test-backupstore-pod
    image: docker.io/erezhorev/dockerized_nfs_server
    securityContext:
      privileged: true
    ports:
    # dummy port to keep k8s happy
    - containerPort: 1111
      name: longhorn-nfs
    args: ["/opt/backupstore"]
---
kind: Service
apiVersion: v1
metadata:
  name: longhorn-test-nfs-svc
spec:
  selector:
    app: longhorn-test-nfs
  clusterIP: None
  ports:
  - name: notnecessary
    port: 1234
    targetPort: 1234
```

**Comment by niusmallnan** [comment]
Final test results:
```
============================= test session starts ==============================
platform linux2 -- Python 2.7.14, pytest-2.9.2, py-1.5.3, pluggy-0.3.1 -- /usr/local/bin/python
cachedir: .cache
rootdir: /integration/tests, inifile:
collecting ... collected 12 items
test_basic.py::test_hosts_and_settings PASSED
test_basic.py::test_volume_basic PASSED
test_basic.py::test_snapshot PASSED
test_basic.py::test_backup PASSED
test_basic.py::test_volume_multinode PASSED
test_driver.py::test_volume_mount PASSED
test_driver.py::test_volume_io PASSED
test_engine_upgrade.py::test_engine_offline_upgrade SKIPPED
test_engine_upgrade.py::test_engine_live_upgrade SKIPPED
test_ha.py::test_ha_simple_recovery PASSED
test_ha.py::test_ha_salvage PASSED
test_recurring_job.py::test_recurring_job PASSED
```

File: csi/controller_server.go [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+		logrus.Errorf("DeleteVolume: invalid delete volume req: %v", req)
+		return nil, err
+	}
+
+	existVol, err := cs.apiClient.Volume.ById(req.GetVolumeId())
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if existVol == nil {
```
Comment (yasker) [inline csi/controller_server.go]: The call should be idempotent: https://github.com/container-storage-interface/spec/blob/master/spec.md#deletevolume

File: csi/controller_server.go [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+		msg := fmt.Sprintf("ControllerPublishVolume: the volume %s not exists", req.GetVolumeId())
+		logrus.Warn(msg)
+		return nil, status.Error(codes.Aborted, msg)
+	}
+	if existVol.State == "attaching" {
+		return nil, status.Errorf(codes.Aborted, "The volume %s is attaching", req.GetVolumeId())
+	}
+
+	needToAttach := false
+	if existVol.State == "detached" {
```
Comment (yasker) [inline csi/controller_server.go]: or `detaching`

File: csi/controller_server.go [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+		logrus.Warn(msg)
+		return nil, status.Error(codes.Aborted, msg)
+	}
+	if existVol.State == "detaching" {
+		return nil, status.Errorf(codes.Aborted, "The volume %s is detaching", req.GetVolumeId())
+	}
+
+	needToDetach := false
+	//TODO Seems we should separate the healthiness of volume from it's attached or detached in the manager API.
+	if existVol.State == "healthy" || existVol.State == "degraded" {
```
Comment (yasker) [inline csi/controller_server.go]: Or `attaching`

File: csi/controller_server.go [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	longhornclient "github.com/rancher/longhorn-manager/client"
+	"github.com/rancher/longhorn-manager/util"
+)
+
+const (
+	oneGB               = 1073741824
+	timeoutAttachDetach = 30 * time.Second
```
Comment (yasker) [inline csi/controller_server.go]: On a heavily loaded machine, attaching may take longer. Make it 120s by default for now.

**Comment by yasker** [comment]
And in the CSI deployment yaml, the longhorn related pods should be named with `longhorn-` prefix, not `csi-` prefix, unless CSI spec required to do so.

File: csi/controller_server.go (hunk 1 above)
Comment (niusmallnan) [reply csi/controller_server.go]: Done

File: csi/controller_server.go (hunk 4 above)
Comment (niusmallnan) [reply csi/controller_server.go]: Done

File: csi/controller_server.go (hunk 2 above)
Comment (niusmallnan) [reply csi/controller_server.go]: Check the volume state, return `Aborted` if `detaching`
```
if existVol.State == string(types.VolumeStateAttaching) || existVol.State == string(types.VolumeStateDetaching) {
    return nil, status.Errorf(codes.Aborted, "The volume %s is %s", req.GetVolumeId(), existVol.State)
}
```

Also return `Aborted` if attach timeout.

**Comment by yasker** [comment]
LGTM. Waiting for the CSI test cases in the longhorn-tests.

**Comment by yasker** [comment]
Merged, thanks.

Please continue to work on the automatic deployment of CSI.

**Pull request merged**